File: internal/model/sales_agent.go
package model

import (
	"time"

	"gorm.io/gorm"
)

// SalesAgent represents an ISP hotspot voucher sales agent.
type SalesAgent struct {
	ID            string         `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	RouterID      string         `gorm:"type:uuid;not null;index" json:"router_id"`
	Name          string         `gorm:"type:varchar(100);not null" json:"name"`
	Phone         *string        `gorm:"type:varchar(20)" json:"phone"`
	Username      string         `gorm:"type:varchar(50);not null;uniqueIndex" json:"username"`
	PasswordHash  string         `gorm:"type:varchar(255);not null" json:"-"`
	Status        string         `gorm:"type:varchar(20);not null;default:'active';check:status IN ('active','inactive')" json:"status"`
	VoucherMode   string         `gorm:"type:varchar(20);not null;default:'mix';check:voucher_mode IN ('mix','num','alp')" json:"voucher_mode"`
	VoucherLength int            `gorm:"type:integer;not null;default:6" json:"voucher_length"`
	VoucherType   string         `gorm:"type:varchar(10);not null;default:'upp';check:voucher_type IN ('upp','up')" json:"voucher_type"`
	BillDiscount  float64        `gorm:"type:decimal(15,2);not null;default:0" json:"bill_discount"`
	BillingCycle  string         `gorm:"type:varchar(20);not null;default:'monthly';check:billing_cycle IN ('weekly','monthly')" json:"billing_cycle"`
	BillingDay    int            `gorm:"type:integer;not null;default:1" json:"billing_day"`
	CreatedAt     time.Time      `gorm:"type:timestamptz;not null;default:CURRENT_TIMESTAMP" json:"created_at"`
	UpdatedAt     time.Time      `gorm:"type:timestamptz;not null;default:CURRENT_TIMESTAMP" json:"updated_at"`
	DeletedAt     gorm.DeletedAt `gorm:"index" json:"-"`

	Router        MikrotikRouter      `gorm:"foreignKey:RouterID"`
	ProfilePrices []SalesProfilePrice `gorm:"foreignKey:SalesAgentID"`
	HotspotSales  []HotspotSale       `gorm:"foreignKey:SalesAgentID"`
}

func (SalesAgent) TableName() string { return "sales_agents" }

// SalesProfilePrice stores per-agent hotspot profile pricing overrides.
type SalesProfilePrice struct {
	ID            string    `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	SalesAgentID  string    `gorm:"type:uuid;not null;index" json:"sales_agent_id"`
	ProfileName   string    `gorm:"type:varchar(100);not null" json:"profile_name"`
	BasePrice     float64   `gorm:"type:decimal(15,2);not null;default:0" json:"base_price"`
	SellingPrice  float64   `gorm:"type:decimal(15,2);not null;default:0" json:"selling_price"`
	VoucherLength *int      `gorm:"type:integer" json:"voucher_length"`
	IsActive      bool      `gorm:"type:boolean;not null;default:true" json:"is_active"`
	CreatedAt     time.Time `gorm:"type:timestamptz;not null;default:CURRENT_TIMESTAMP" json:"created_at"`

	SalesAgent SalesAgent `gorm:"foreignKey:SalesAgentID"`
}

func (SalesProfilePrice) TableName() string { return "sales_profile_prices" }

File: web/src/lib/schemas/sales-agent.ts
import { z } from 'zod'
import { ApiResponseSchema, MessageResponseSchema as _msg } from '@/lib/schemas/auth'
import { MetaSchema } from '@/lib/schemas/router'

export const SalesAgentResponseSchema = z.object({
  id: z.string().uuid(),
  router_id: z.string().uuid(),
  name: z.string(),
  phone: z.string().nullish(),
  username: z.string(),
  status: z.enum(['active', 'inactive']),
  voucher_mode: z.enum(['mix', 'num', 'alp']).nullish(),
  voucher_length: z.number().int().nullish(),
  voucher_type: z.enum(['upp', 'up']).nullish(),
  bill_discount: z.number().nullish(),
  billing_cycle: z.enum(['weekly', 'monthly']).nullish(),
  billing_day: z.number().int().nullish(),
  created_at: z.string(),
  updated_at: z.string(),
})

export type SalesAgentResponse = z.infer<typeof SalesAgentResponseSchema>

export const CreateSalesAgentRequestSchema = z.object({
  router_id: z.string().uuid(),
  name: z.string().min(1),
  username: z.string().min(1),
  password: z.string().min(6),
  phone: z.string().optional(),
  status: z.enum(['active', 'inactive']).optional(),
  voucher_mode: z.enum(['mix', 'num', 'alp']).optional(),
  voucher_length: z.number().int().optional(),
  voucher_type: z.enum(['upp', 'up']).optional(),
  bill_discount: z.number().optional(),
  billing_cycle: z.enum(['weekly', 'monthly']).optional(),
  billing_day: z.number().int().min(1).max(31).optional(),
})

export type CreateSalesAgentRequest = z.infer<typeof CreateSalesAgentRequestSchema>

export const UpdateSalesAgentRequestSchema = z.object({
  name: z.string().optional(),
  phone: z.string().optional(),
  password: z.string().optional(),
  status: z.enum(['active', 'inactive']).optional(),
  voucher_mode: z.enum(['mix', 'num', 'alp']).optional(),
  voucher_length: z.number().int().optional(),
  voucher_type: z.enum(['upp', 'up']).optional(),
  bill_discount: z.number().optional(),
  billing_cycle: z.enum(['weekly', 'monthly']).optional(),
  billing_day: z.number().int().min(1).max(31).optional(),
})

export type UpdateSalesAgentRequest = z.infer<typeof UpdateSalesAgentRequestSchema>

export const SalesProfilePriceResponseSchema = z.object({
  id: z.string().uuid(),
  sales_agent_id: z.string().uuid(),
  profile_name: z.string(),
  base_price: z.number(),
  selling_price: z.number(),
  voucher_length: z.number().int().nullish(),
  is_active: z.boolean(),
  created_at: z.string(),
})

export type SalesProfilePriceResponse = z.infer<typeof SalesProfilePriceResponseSchema>

export const UpsertProfilePriceRequestSchema = z.object({
  base_price: z.number().optional(),
  selling_price: z.number().optional(),
  voucher_length: z.number().int().optional(),
  is_active: z.boolean().optional(),
})

export type UpsertProfilePriceRequest = z.infer<typeof UpsertProfilePriceRequestSchema>

// Response wrappers
export const SalesAgentListResponseSchema = ApiResponseSchema(z.array(SalesAgentResponseSchema)).extend({
  meta: MetaSchema.optional(),
})

export const SalesAgentDetailResponseSchema = ApiResponseSchema(SalesAgentResponseSchema)

export const SalesProfilePriceListResponseSchema = ApiResponseSchema(z.array(SalesProfilePriceResponseSchema))

export const SalesProfilePriceDetailResponseSchema = ApiResponseSchema(SalesProfilePriceResponseSchema)

// Generate agent invoice
export const GenerateAgentInvoiceRequestSchema = z.object({
  period_start: z.string(),
  period_end: z.string(),
})

export type GenerateAgentInvoiceRequest = z.infer<typeof GenerateAgentInvoiceRequestSchema>

export const MarkPaidRequestSchema = z.object({
  paid_amount: z.number().min(0),
})
